binding/c: fix misc validation bugs

hzhou opened the pull request:
## Pull Request Description

These are corner cases that the binding generation layer either missed or over-checked. They are discovered by `mpi4py`'s testsuite.

Fixes #5159 #5149

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou commented:
With this PR, the status of `mpi4py`'s testsuite:
```
$ python runtests.py
[0@Tiger] Python 3.6 (/usr/bin/python)
[0@Tiger] MPI 3.1 (MPICH 4.0.0)
[0@Tiger] mpi4py 3.1.0a0 (/home/hzhou/opt/mpi4py/lib/python3.6/site-packages/mpi4py)
...............................................................................................................................................................................................................................................................................................................ssssssssssssssssssssssss..........................................................ss...................................................................ssssssssssss............................................................ssss............s........................................................................................................................................................................................s.............................s...ss..s....s......sssssssss.................sssssssss...........ssssss............ssssss....s..s.........................................................................................................................................................................s...................................................................ssssssssssssssssssssssssssssssssssssssss...............................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 1355 tests in 21.347s

OK (skipped=122)
```

dalcinl commented:
@hzhou About this one:

```
src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c:372: MPIR_Typerep_pack_external: Assertion `0' failed.
```

Would it be too much trouble for you to use the usual MPI error handling mechanisms?

dalcinl commented:
@hzhou I just found another bug thanks to the pack/unpack external tests:

Look  at the file `src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c`, and search the following block of code (line 313):

```c
#ifdef HAVE_FORTRAN_BINDING
        case MPI_FLOAT:
        case MPI_REAL:
        case MPI_REAL4:
            PACK_EXTERNAL_FLOAT_MAPPED(iov, outbuf, max_iov_len, basic_type_size,
                                       typerep_float32_t);
            break;
#endif /* HAVE_FORTRAN_BINDING */
```

`MPI_FLOAT` is not a Fortran datatype. You probably want to move this case below, right above the `MPI_DOUBLE` case?

dalcinl commented:
@hzhou Did you run the testsuite in parallel? Some tests are skipped if you run in just one process. Also, some tests are are skipped if yo do not use `mpiexec` (the singleton MPI feature from 3340)

Please try running the following way:

```shell
mpiexec -n 4  python test/runtests.py
```

PS: Could you share with me what's the configure line you are using? What PM are you using? The old hydra or the new hydra2? I'm running your branch, and I still have failures with accept/connect and spawn tests.

dalcinl commented:
@hzhou Did you run the testsuite in parallel? Some tests are skipped if you run in just one process. Also, some tests are are skipped if yo do not use mpiexec (the singleton MPI feature from #3340)

Please try running the following way:

```
mpiexec -n 4  python test/runtests.py
```

PS: Could you share with me what's the configure line you are using? What PM are you using? The old hydra or the new hydra2? I'm running your branch, and I still have failures with accept/connect and spawn tests.

dalcinl commented:
I pushed testsuite fixes to mpi4py branch [testing/mpich](/mpi4py/mpi4py/tree/testing/mpich). 

There are some remaining issues on my Fedora 33 wokstation:

1. MPICH's pack/unpack external support is broken for `MPI_LONG`. The MPI standard mandates to use size=4, but `sizeof(long)` is 8 on the LP64 model.
2. I get failures when using accept/connect and spawn. I did not pass `--with-pm` to configure, so I using the default (old hydra?).

hzhou commented:
> 
> 
> @hzhou About this one:
> 
> ```
> src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c:372: MPIR_Typerep_pack_external: Assertion `0' failed.
> ```
> 
> Would it be too much trouble for you to use the usual MPI error handling mechanisms?

I agree that we should turn the assertion into error. But my initial attempt ended up in a dead loop, so I decided to kick the can down to the next block :)

hzhou commented:
> 
> 
> @hzhou Did you run the testsuite in parallel? Some tests are skipped if you run in just one process. Also, some tests are are skipped if yo do not use mpiexec (the singleton MPI feature from #3340)
> 
> Please try running the following way:
> 
> ```
> mpiexec -n 4  python test/runtests.py
> ```
> 
> PS: Could you share with me what's the configure line you are using? What PM are you using? The old hydra or the new hydra2? I'm running your branch, and I still have failures with accept/connect and spawn tests.

I suspected that :) but unconsciously choose to ignore that suspicion to feel more accomplished for the day. Will work on it.

hzhou commented:
> 
> 
> @hzhou I just found another bug thanks to the pack/unpack external tests:
> 
> Look at the file `src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c`, and search the following block of code (line 313):
> 
> ```c
> #ifdef HAVE_FORTRAN_BINDING
>         case MPI_FLOAT:
>         case MPI_REAL:
>         case MPI_REAL4:
>             PACK_EXTERNAL_FLOAT_MAPPED(iov, outbuf, max_iov_len, basic_type_size,
>                                        typerep_float32_t);
>             break;
> #endif /* HAVE_FORTRAN_BINDING */
> ```
> 
> `MPI_FLOAT` is not a Fortran datatype. ~You probably want to move this case below, right above the `MPI_DOUBLE` case?~
> 
> I believe this is the proper fix:
> 
> ```diff
> diff --git a/src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c b/src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c
> index 0a726cbab..27c4e0a4b 100644
> --- a/src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c
> +++ b/src/mpi/datatype/typerep/src/typerep_yaksa_pack_external.c
> @@ -310,14 +310,14 @@ int MPIR_Typerep_pack_external(const void *inbuf, MPI_Aint incount, MPI_Datatype
>              PACK_EXTERNAL_INT_MAPPED(iov, outbuf, max_iov_len, basic_type_size, int64_t);
>              break;
>  
> -#ifdef HAVE_FORTRAN_BINDING
>          case MPI_FLOAT:
> +#ifdef HAVE_FORTRAN_BINDING
>          case MPI_REAL:
>          case MPI_REAL4:
> +#endif /* HAVE_FORTRAN_BINDING */
>              PACK_EXTERNAL_FLOAT_MAPPED(iov, outbuf, max_iov_len, basic_type_size,
>                                         typerep_float32_t);
>              break;
> -#endif /* HAVE_FORTRAN_BINDING */
>  
>          case MPI_DOUBLE:
>          case MPI_C_COMPLEX:
> @@ -487,14 +487,14 @@ int MPIR_Typerep_unpack_external(const void *inbuf, void *outbuf, MPI_Aint outco
>              UNPACK_EXTERNAL_INT_MAPPED(inbuf, iov, max_iov_len, basic_type_size, int64_t);
>              break;
>  
> -#ifdef HAVE_FORTRAN_BINDING
>          case MPI_FLOAT:
> +#ifdef HAVE_FORTRAN_BINDING
>          case MPI_REAL:
>          case MPI_REAL4:
> +#endif /* HAVE_FORTRAN_BINDING */
>              UNPACK_EXTERNAL_FLOAT_MAPPED(inbuf, iov, max_iov_len, basic_type_size,
>                                           typerep_float32_t);
>              break;
> -#endif /* HAVE_FORTRAN_BINDING */
>  
>          case MPI_DOUBLE:
>          case MPI_C_COMPLEX:
> ```

Thanks! I'll try that.

hzhou commented:
Added more fixes, now it should pass the tests with `mpirun -n 4 python runtests.py`.

hzhou commented:
@dalcinl 

> I pushed testsuite fixes to mpi4py branch [testing/mpich](/mpi4py/mpi4py/tree/testing/mpich).
> 

I have fixed the `pack_external` (applied your patch and added error return), so the only patch needed is the `datatype == MPI_DATATYPE_NULL` condition.

> There are some remaining issues on my Fedora 33 wokstation:
> 
>     1. MPICH's pack/unpack external support is broken for `MPI_LONG`. The MPI standard mandates to use size=4, but `sizeof(long)` is 8 on the LP64 model.

Added to my TODO

> 
>     2. I get failures when using accept/connect and spawn. I did not pass `--with-pm` to configure, so I'm using the default (old hydra?).

The spawn bugs are fixed. `hydra` is fine. I think we are going to give up `hydra2` and focus on bring `hydra` to exascale.

hzhou commented:
>     1. MPICH's pack/unpack external support is broken for `MPI_LONG`. The MPI standard mandates to use size=4, but `sizeof(long)` is 8 on the LP64 model.

I don't see how the code is broken. We group `MPI_LONG` with e.g. `MPI_INT32_T`,  thus are treating it with 4 bytes. Could you describe a reproducer?

hzhou commented:
test:mpich/ch4/most

hzhou commented:
test:mpich/ch3/most

dalcinl commented:
A reproducer for the issues with pack/unpack `MPI_LONG`, just put it in a script and run it.

```python
import sys
from mpi4py import MPI
from array import array

try:
    typecode = sys.argv[1]
except IndexError:
    typecode = 'l' # long

datatype = MPI._typedict[typecode]
print(datatype.Get_name())

iarray = array(typecode, [1,2,3])
oarray = array(typecode, [0]) * len(iarray)

n = datatype.Pack_external_size("external32", len(iarray))
print(n, n//len(iarray))

work = bytearray(n)

p = 0
p = datatype.Pack_external("external32", iarray, work, p)
print(p, p//len(iarray))

q = 0
q = datatype.Unpack_external("external32", work, q, oarray)
print(q, q//len(iarray))

print(oarray)
assert oarray == iarray
```

By default it will test `long` (typecode `"l"`), but you can pass other [typecodes](https://docs.python.org/3/library/array.html) in the command line.

```
$ python script.py i
MPI_INT
12 4
12 4
12 4
array('i', [1, 2, 3])

$ python script.py l
MPI_LONG
12 4
24 8
24 8
array('l', [12884901889, 2714902768, 0])
Traceback (most recent call last):
  File "/home/dalcinl/Devel/mpi4py-dev/tmp-4.py", line 30, in <module>
    assert oarray == iarray
AssertionError
```

Valgrind clearly shows problems:

```
$ valgrind -q python script.py 
MPI_LONG
12 4
24 8
==983150== Invalid read of size 8
==983150==    at 0x13E0BBEB: MPIR_Typerep_unpack_external (typerep_yaksa_pack_external.c:477)
==983150==    by 0x13E10D4C: MPIR_Unpack_external_impl (datatype_impl.c:99)
==983150==    by 0x13877615: internal_Unpack_external (unpack_external.c:75)
==983150==    by 0x138778C6: PMPI_Unpack_external (unpack_external.c:160)
==983150==    by 0x134D2055: __pyx_pf_6mpi4py_3MPI_8Datatype_64Unpack_external (mpi4py.MPI.c:80589)
==983150==    by 0x134D1D03: __pyx_pw_6mpi4py_3MPI_8Datatype_65Unpack_external (mpi4py.MPI.c:80460)
==983150==    by 0x498A794: ??? (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x497C687: _PyEval_EvalFrameDefault (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x497B090: ??? (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49F9D94: _PyEval_EvalCodeWithName (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49F9D2C: PyEval_EvalCodeEx (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49F9CDE: PyEval_EvalCode (in /usr/lib64/libpython3.9.so.1.0)
==983150==  Address 0x195a1b10 is 3 bytes after a block of size 13 alloc'd
==983150==    at 0x4839809: malloc (vg_replace_malloc.c:307)
==983150==    by 0x49F53E5: ??? (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x498FAB5: PyObject_Realloc (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49CD9EB: PyByteArray_Resize (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49B6A33: ??? (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x4984FF2: ??? (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x4984D22: _PyObject_MakeTpCall (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49811AB: _PyEval_EvalFrameDefault (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x497B090: ??? (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49F9D94: _PyEval_EvalCodeWithName (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49F9D2C: PyEval_EvalCodeEx (in /usr/lib64/libpython3.9.so.1.0)
==983150==    by 0x49F9CDE: PyEval_EvalCode (in /usr/lib64/libpython3.9.so.1.0)
==983150== 
24 8
array('l', [12884901889, 0, 0])
Traceback (most recent call last):
  File "/home/dalcinl/Devel/mpi4py-dev/tmp-4.py", line 30, in <module>
    assert oarray == iarray
AssertionError
```

dalcinl commented:
@hzhou I found the problem, it is just a bit of confusing about what are the src/dest types in unpack, plus wrong computation of the number of packed/unpacked bytes. 

Look at the attached patch. It contains the minimal changes required to make things work and for you to quickly see where the problems are. However, you should do better, I'm setting the num bytes variables within the pack/unpack macros without passing them as arguments.

Attachment: [patch.txt](https://github.com/pmodels/mpich/files/6177059/patch.txt)

dalcinl commented:
> The spawn bugs are fixed.

I think there are still some problems. I've setup GitHub Actions to test MPICH+mpi4py.
https://github.com/mpi4py/mpi4py-testing/runs/2159624157
I had to cancel the workflow run.

1. The `hydra` + `ch4:ofi` build got stuck in a test that uses spawn. The failure is hard to reproduce on a modern workstation/laptop. I can reproduce the deadlock on a Ubuntu 20.04 virtual machine with 2 CPU cores (which is the setup of GitHub-hosted runners). FYI, this test runs in one MPI process that calls spawn many times, in some cases asking for one child, and in some other cases for two children. I have no clue how to further debug the issue. Perhaps it is mpi4py's fault, but it is quite suspicious that the number of processor cores has an effect.
2. Note the gforker failures: `[server]: unknown cmd publish_name`. I guess this is expected, after all this all based on `fork()`, right? Any tips on how to "detect" this case at runtime and then skip tests? Perhaps there is a quick fix that would go through the normal MPI error handling machinery? Or should I just refrain from testing with gforker ?

hzhou commented:
> 
> 
> @hzhou I found the problem, it is just a bit of confusion about what are the src/dest types in unpack, plus wrong computation of the number of packed/unpacked bytes.
> 
> Look at the attached patch. It contains the minimal changes required to make things work and for you to quickly see where the problems are. However, you should do better, I'm setting the num bytes variables within the pack/unpack macros without passing them as arguments.
> 
> Attachment: [patch.txt](https://github.com/pmodels/mpich/files/6177059/patch.txt)

Thanks! I'll apply your fix in a separate PR and add a test case for that.

hzhou commented:
test:mpich/ch4/most

hzhou commented:
> 
> 
> > @hzhou I found the problem, it is just a bit of confusion about what are the src/dest types in unpack, plus wrong computation of the number of packed/unpacked bytes.
> > Look at the attached patch. It contains the minimal changes required to make things work and for you to quickly see where the problems are. However, you should do better, I'm setting the num bytes variables within the pack/unpack macros without passing them as arguments.
> > Attachment: [patch.txt](https://github.com/pmodels/mpich/files/6177059/patch.txt)
> 
> Thanks! I'll apply your fix in a separate PR and add a test case for that.

Hmm, looks like there are more issues to the code. It ignores the `big-endian` packing requirement. In addition, the packed data may be out-of-alignment, thus should not be accessed as integer or float.

dalcinl commented:
> It ignores the `big-endian` packing requirement.

Ups! I somehow naively assumed that Yaksa was taking care of endianess. But now that I look better, I realize that's not the case. That should be easy to fix. The alignment issue, however, is something I have no idea how to implement efficiently, I never did it.

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/ucx

dalcinl commented:
My `ch4:ofi` [logs](https://github.com/mpi4py/mpi4py-testing/runs/2163948903) are flooded with the following warning, and I believe this is affecting the compilation time on the CI VMs.

```
./src/mpi/coll/algorithms/common/algo_common.h:54:76: note: format string is defined here
   54 |     MPL_DBG_MSG_FMT(MPIR_DBG_COLL, VERBOSE, (MPL_DBG_FDEST, "num_segments %d", *num_segments));
      |                                                                           ~^
      |                                                                            |
      |                                                                            int
      |                                                                           %ld
```

Using `"%llu"` format and adding a cast to `(unsigned long long)` should be enough, right? Could you please push a fix to this PR?

dalcinl commented:
I have a simple reproducer for [spawn-related deadlocks](https://github.com/mpi4py/mpi4py-testing/runs/2163948903) when using `ch4:ofi`

First, put this code in a script:
```python
from mpi4py import MPI
import sys

def main_parent():
    comm = MPI.COMM_SELF.Spawn(
        sys.executable,
        [__file__, '--child'],
        maxprocs=2,
    )
    buf = b'@'
    MPI.Request.Waitall([
        comm.Issend([buf, 1, MPI.BYTE], dest=pid, tag=0)
        for pid in range(comm.Get_remote_size())
    ])
    comm.Disconnect()


def main_child():
    comm = MPI.Comm.Get_parent()
    buf = bytearray(1)
    comm.Recv([buf, 1, MPI.BYTE], source=0, tag=0)
    assert buf == b'@'
    comm.Disconnect()


if '--child' in sys.argv:
    main_child()
else:
    main_parent()
```

Next, disable CPU cores to have only two online (to revert, use `echo 1`):
```bash
sudo su -l
for i in `seq 2 $(($(nproc)-1))`; do echo 0 > /sys/devices/system/cpu/cpu$i/online; done
cat /sys/devices/system/cpu/online
```

Next, run the script with one MPI process:
```
mpiexec -n 1 python3 script.py
```
It should deadlock. Note that this example involves a total of three MPI processes (1 parent + 2 childen). As you have only two CPU cores online, this qualifies as oversubscription, right?

Finally, edit the script and replace `comm.Issend(...)` with `comm.Isend(...)` and run it again.

Now It should work! So perhaps this is not really related to spawn, but some progress bug with synchronous-mode sends in `ch4:ofi` when you are oversubscribed on low CPU counts.

hzhou commented:
> 
> 
> My `ch4:ofi` [logs](https://github.com/mpi4py/mpi4py-testing/runs/2163948903) are flooded with the following warning, and I believe this is affecting the compilation time on the CI VMs.
> 
> ```
> ./src/mpi/coll/algorithms/common/algo_common.h:54:76: note: format string is defined here
>    54 |     MPL_DBG_MSG_FMT(MPIR_DBG_COLL, VERBOSE, (MPL_DBG_FDEST, "num_segments %d", *num_segments));
>       |                                                                           ~^
>       |                                                                            |
>       |                                                                            int
>       |                                                                           %ld
> ```
> 
> Using `"%llu"` format and adding a cast to `(unsigned long long)` should be enough, right? Could you please push a fix to this PR?

I hate those debug lines -- I want to delete them all, but some in our team are adamant about keeping them :( -- Anyway, I think I'll do this fix separately (potentially argue with my team one more time 😄 ).

raffenet commented on maint/local_python/binding_c.py:
```diff
@@ -1801,7 +1801,7 @@ def dump_validation(func, t):
             # allow MPI_ANY_SOURCE, MPI_PROC_NULL
             G.out.append("MPIR_ERRTEST_RECV_RANK(%s, %s, mpi_errno);" % (comm_ptr, name))
-        elif RE.match(r'(dest|target_rank)$', name) or RE.match(r'mpi_win_', func_name, re.IGNORECASE):
+        elif RE.match(r'(source|dest|target_rank)$', name) or RE.match(r'mpi_win_', func_name, re.IGNORECASE):
```
Does this allow `MPI_ANY_SOURCE`?

hzhou replied:
For the functions that allow `MPI_ANY_SOURCE`, it should be captured by the first if-branch. If it leaks to here, then we need fix (the first condition).

I am not sure whether `sendrecv` accepts `MPI_ANY_SOURCE` ...

raffenet replied:
MPI_Sendrecv contains a recv operation. It allows `MPI_ANY_SOURCE`.

raffenet replied:
What I mean is, the receive in `MPI_Sendrecv` is no different than a normal `MPI_Recv`. It just happens concurrently with the send.

hzhou replied:
done.

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most

raffenet commented on src/binding/c/op_api.txt:
```diff
@@ -20,6 +20,13 @@ MPI_Op_create:
 MPI_Op_free:
     .desc: Frees a user-defined combination function handle
+{ -- error_check --
+    if (op_ptr->kind < MPIR_OP_KIND__USER_NONCOMMUTE) {
```
This is nitpicking, but would `if (HANDLE_IS_BUILTIN(*op)` be somewhat more robust?

hzhou replied:
Sounds good. Fixed.

raffenet commented on maint/local_python/binding_c.py:
```diff
@@ -1801,6 +1801,8 @@ def dump_validation(func, t):
         if RE.match(r'mpi_(i?m?probe|i?recv|recv_init)$', func_name, re.IGNORECASE) or name == "recvtag":
             # allow MPI_ANY_SOURCE, MPI_PROC_NULL
             G.out.append("MPIR_ERRTEST_RECV_RANK(%s, %s, mpi_errno);" % (comm_ptr, name))
+        elif RE.match(r'mpi_(isendrecv)', func_name, re.IGNORECASE) and name == "source":
```
This won't catch `MPI_Sendrecv` or `MPI_Sendrecv_replace`, will it? Should it be something like `r'mpix?_(i?sendrecv)'`?

hzhou replied:
Yes. I don't what I was thinking :) 

Fixed.

We don't need `mpix?_` because the `mpix` prefix has not been added here yet (it is added in `get_function_name`).

raffenet reviewed (changes requested):
A couple minor issues I noticed. Can we split out the ch4 changes into another PR? I haven't gone over them closely yet.

hzhou commented:
>  Can we split out the ch4 changes into another PR?

Sure.

hzhou commented:
test:mpich/spellchecker

hzhou commented:
test:mpich/spellcheck

hzhou commented:
test:mpich/ch4/ofi

hzhou commented:
test:mpich/custom
netmod: ch3

hzhou commented:
test:mpich/custom
netmod: ch3:tcp
